go/internal/feast/sqliteonlinestore.go

```diff
@@ -0,0 +1,147 @@
+package feast
+
+import (
+	"database/sql"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	"context"
+	"fmt"
+
+	"github.com/feast-dev/feast/go/protos/feast/serving"
+	"github.com/feast-dev/feast/go/protos/feast/types"
+	_ "github.com/mattn/go-sqlite3"
+	"google.golang.org/protobuf/proto"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type SqliteOnlineStore struct {
+	// Feast project name
+	project string
+	path    string
+	db      *sql.DB
+	db_mu   sync.Mutex
+}
+
+// Creates a new sqlite online store object. onlineStoreConfig should have relative path of database file with respect to repoConfig.repoPath.
+func NewSqliteOnlineStore(project string, repoConfig *RepoConfig, onlineStoreConfig map[string]interface{}) (*SqliteOnlineStore, error) {
+	store := SqliteOnlineStore{project: project}
+	if db_path, ok := onlineStoreConfig["path"]; !ok {
+		return nil, fmt.Errorf("cannot find sqlite path %s", db_path)
+	} else {
+		if dbPathStr, ok := db_path.(string); !ok {
+			return nil, fmt.Errorf("cannot find convert sqlite path to string %s", db_path)
+		} else {
+			store.path = fmt.Sprintf("%s/%s", repoConfig.RepoPath, dbPathStr)
+			db, err := initializeConnection(store.path)
+			if err != nil {
+				return nil, err
+			}
+			store.db = db
+		}
+	}
+
+	return &store, nil
+}
+
+func (s *SqliteOnlineStore) Destruct() {
+	s.db.Close()
+}
+
+// Returns FeatureData 2D array. Each row corresponds to one entity value and each column corresponds to a single feature where the number of columns should be
+// same length as the length of featureNames. Reads from every table in featureViewNames with the entity keys described.
+func (s *SqliteOnlineStore) OnlineRead(ctx context.Context, entityKeys []types.EntityKey, featureViewNames []string, featureNames []string) ([][]FeatureData, error) {
+	featureCount := len(featureNames)
+	_, err := s.getConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	project := s.project
+	results := make([][]FeatureData, len(entityKeys))
+	entityNameToEntityIndex := make(map[string]int)
+	in_query := make([]string, len(entityKeys))
+	serialized_entities := make([]interface{}, len(entityKeys))
+	for i := 0; i < len(entityKeys); i++ {
+		serKey, err := serializeEntityKey(&entityKeys[i])
+		if err != nil {
+			return nil, err
+		}
+		// TODO: fix this, string conversion is not safe
+		entityNameToEntityIndex[string(*serKey)] = i
+		// for IN clause in read query
+		in_query[i] = "?"
+		serialized_entities[i] = *serKey
+	}
+	featureNamesToIdx := make(map[string]int)
+	for idx, name := range featureNames {
+		featureNamesToIdx[name] = idx
+	}
+
+	for idx := 0; idx < len(entityKeys); idx++ {
+		results[idx] = make([]FeatureData, featureCount)
+	}
+	for _, featureViewName := range featureViewNames {
```

Comment:
do we need this for loop here? Can't we just query all keys in one go?

Reply:
Oh, I see you have table per feature view. I guess that's fine for sqlite implementation.

Reply:
Yes, I am adhering to the API set by the online store API.